Add support for permitlisten to authorized_keys

mitar opened the pull request:
Fixes #382.

mitar committed:
Add support for permitlisten to authorized_keys

mkj commented:
Ran OK now

melnikovsky commented:
Does the semantics of proposed `permitlisten` really agree with that from OpenSSH?

> SSHD(8)   permitlisten="[host:]port"
> Limit remote port forwarding with the ssh(1) -R option such that it may only listen on the specified host (optional) and port.

It seems that

> DROPBEAR(8)   permitlisten="[host:]port"
> Restrict remote port forwarding so that connection is allowed only from the specified host and port.

sounds more like `permit-accept-from`. Am I missing something?

mitar commented:
Hm, I think documentation I added is then bad. The code checks the address and port which the user provided over SSH connection and is address and port on which to listen.

melnikovsky commented:
OK then, thanks. I only reviewed the docs.
May I recommend to change the wording in the code as well.

mitar commented:
Yes, feel free to suggest better wording. You can make a PR against my branch.

mkj committed:
Reword permitlisten to avoid "source"

The bind/listen address is what should be used. "Source" could be
confused with the originating TCP host connecting to a listener.

mkj commented:
I've rebased this and fixed the wording.

mitar commented:
Awesome, thanks!

mkj committed:
Only accept a port for permitlisten

For the time being a plain permitlisten="port" is adequate.

Host support is more complicated with various special hostnames of "*"
or "localhost", "127.0.0.1", "::1", etc. There are also interactions
with "-a"/GatewayPorts.

mkj commented:
Thanks!

mitar commented:
Thanks!

I am curious, what is the reason for removing restricting host? Because I think it is not really important what is the (special) meaning behind the host string, because admin sets what it is allowed and clients then simply can or cannot send/use that string. So the meaning of allowed host string is left to the admin to decide.

mkj commented:
At the moment I don't think the host support was ready, and plain port by itself seems useful and still compatible with OpenSSH syntax. Better to get basic support merged before the next release, `host:` could be added later if it's going to be used.

An empty host `-R :bindport:fwdhost:fwdport` is valid to be sent by the client for "listen on any" (if Dropbear is running with `-a`), the PR didn't handle that (https://www.rfc-editor.org/rfc/rfc4254#page-17). Supporting port of `*` probably makes sense when combined with host restrictions.

Dropbear's `split_address_port()` has kind of inconsistent semantics, it could be improved but other callers of it would also need to be fixed.

mitar commented:
Thanks for explanation.
